File: cpp/procedural/pid.cc
#include <iostream>

int main() {
    const double M = 2.0;       // Mass
    const double P_i = 0.0;     // Initial position
    const double P_t = 10.0;    // Target position
    const double k_p = 0.5, k_i = 0.1, k_d = 0.2;
    const double dt = 0.1;      // Time interval
    const int s = 1000;          // Steps

    double position = P_i;
    double velocity = 0.0;
    double integral = 0.0;
    double prev_error = 0.0;

    for (int step = 0; step < s; ++step) {
        double error = P_t - position;
        integral += error * dt;
        double derivative = (error - prev_error) / dt;
        double force = k_p * error + k_i * integral + k_d * derivative;

        double acceleration = force / M;
        velocity += acceleration * dt;
        position += velocity * dt;
        prev_error = error;

        // std::cout << "Step " << step + 1 << ": Position = " << position << std::endl;
    }

    return 0;
}

File: go/procedural/pid.go
package main

import "fmt"

type State struct {
	Position  float64
	Velocity  float64
	Integral  float64
	PrevError float64
}

func simulatePID(current State, target, mass, kP, kI, kD, dt float64, steps int) State {
	for step := 0; step < steps; step++ {
		error := target - current.Position
		derivative := (error - current.PrevError) / dt
		force := kP*error + kI*current.Integral + kD*derivative

		acceleration := force / mass
		current.Velocity += acceleration * dt
		current.Position += current.Velocity * dt
		current.Integral += error * dt
		current.PrevError = error

		// fmt.Printf("Step %d: Position = %.2f\n", step+1, current.Position)
	}
	return current
}

func main() {
	initial := State{Position: 0.0}
	params := map[string]float64{
		"mass":   2.0,
		"target": 10.0,
		"kP":     0.5,
		"kI":     0.1,
		"kD":     0.2,
		"dt":     0.1,
		"steps":  1000,
	}

	final := simulatePID(initial, params["target"], params["mass"],
		params["kP"], params["kI"], params["kD"], params["dt"], int(params["steps"]))
	fmt.Printf("Final position: %.2f\n", final.Position)
}
